packages/components/src/Input/amountInput.tsx

```diff
@@ -0,0 +1,119 @@
+import BigNumber from "bignumber.js";
+import { ChangeEventHandler, ComponentProps, useState } from "react";
+
+import { Result } from "@namada/utils";
+
+import { Input } from "./input";
+import { InputVariants } from "./types";
+
+type BigNumberElement = Omit<HTMLInputElement, "value"> & {
+  value?: BigNumber;
+};
+
+type Props = Omit<ComponentProps<typeof Input>, "value" | "onChange"> & {
+  value?: BigNumber;
+  onChange?: ChangeEventHandler<BigNumberElement>;
+  maxDecimalPlaces?: number;
+};
+
+type ValidationError =
+  | "NotANumber"
+  | "TooSmall"
+  | "TooBig"
+  | "TooManyDecimalPlaces";
+
+const validate = (
+  stringValue: string,
+  props: Pick<Props, "min" | "max" | "maxDecimalPlaces">
+): Result<BigNumber, ValidationError> => {
+  const { min, max, maxDecimalPlaces } = props;
+  const asBigNumber = new BigNumber(stringValue);
+
+  if (asBigNumber.isNaN()) {
+    return Result.err("NotANumber");
+  } else {
+    const decimalPlaces = asBigNumber.decimalPlaces();
+
+    if (decimalPlaces === null) {
+      return Result.err("NotANumber");
+    } else if (
+      maxDecimalPlaces !== undefined &&
+      decimalPlaces > maxDecimalPlaces
+    ) {
+      return Result.err("TooManyDecimalPlaces");
+    } else if (min !== undefined && asBigNumber.isLessThan(min)) {
+      return Result.err("TooSmall");
+    } else if (max !== undefined && asBigNumber.isGreaterThan(max)) {
+      return Result.err("TooBig");
+    } else {
+      return Result.ok(asBigNumber);
```

Comment:
This is just a random thought, but I'm curious if mapping `ifs` predicates to Options like in this: https://gcanti.github.io/fp-ts/modules/Option.ts.html and then doing some kind of matching would reduce number of returns here :)